Dispatch for unary operations

gmarkall opened the pull request:
Dispatch for unary operations is modelled on the dispatch for binary operations.

Its use is demonstrated with the implementation of NEGATE materialisation which is tested from both Java and C++.

This PR contains some of the removal of code that prints out trees as the PR for MAT-228, also to quieten the output down now that we're beginning to execute trees rather than just looking at them.

Buildbot pass: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/385
Librdag coverage: 90.3% (2% higher than current master)
jshim coverage: 20.6% (equal to current master)

gmarkall committed:
[MAT-270] Unary node dispatch implementation.

The Unary node dispatch implementation uses an example of a Negate
runner as a first example.

The negate runner is tested through a Java materialiser test.

gmarkall committed:
[MAT-270] Add C++ test of NEGATE materialisation.

stuartarchibald commented on include/dispatch.hh:
```diff
@@ -53,9 +54,120 @@ public:
+    // impl convert and run for type OGIntegerScalar
+  };
+
+  virtual T run(RegContainer SUPPRESS_UNUSED * reg, OGRealMatrix const SUPPRESS_UNUSED * arg) const
```
should this be pure virtual as in the bin-op case such that implementers of the interface are required to provide "backstop/default" implementations so that if the operation is valid something is guaranteed to run?

gmarkall replied:
Yes, good point. Will make this change.

stuartarchibald replied:
Think same applies to the OGComplexMatrix variant.

gmarkall replied:
I agree.

gmarkall committed:
[MAT-270] Add Java NEGATE materialise test.

gmarkall commented:
Just added the Java-side test that I previously forgot to commit.

Buildbot: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/386

gmarkall committed:
[MAT-270] Force implementation of unary backstops.

Unary backstop methods for Real and Complex matrices must now be
implemented as the base class methods are virtual.

An implementation of these methods for NEGATE is provided, but
they are as yet untested as incorporating them in to the existing
test will require the equals method.

gmarkall commented:
The buildbot pass for the latest push is here: http://devmaths-lx-1:8011/builders/nyqwk2-testing/builds/390

stuartarchibald commented:
Code matches dispatch logic for binary ops. Happy to merge as WIP. Tickets issued to address other known problems (e.g. large trees running from java hanging most likely due to jvm frame overflow).
